### PR TITLE
Enable -Xfatal-warnings and other linty flags

### LINKED ISSUE #648: Scaladoc has broken external links.
At least some external links in scaladoc are broken. See [Request](http://http4s.org/api/0.15/#org.http4s.Response) and click on `Process` from the signature for `bodyAsText` and it craps out because no such host. I couldn't find a declaration anywhere in the build that associates streams with that host, but I didn't look very hard.

### PR DESCRIPTION
**Pull request by rossabaker**
The list of scalac options is stolen from Cats.

We lose two bits of functionality:
1. Can no longer set reason phrase on servlet backends.  It was deprecated, and doesn't work on Tomcat anyway without special server configuration.  I'd rather have fatal warnings than this obscure feature.
2. We lost the Java 8 optimizations for the Scala 2.11 build.

Fixes #648.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -389,22 +391,33 @@ lazy val commonSettings = Seq(
...
+    "-Xfatal-warnings",
+    "-Xlint",
+    "-Yinline-warnings",
+    "-Yno-adapted-args",
+    "-Ywarn-dead-code",
+    "-Ywarn-numeric-widen",
+    "-Ywarn-value-discard",
+    "-Xfuture"
+  ),
+  /* disabled because SI-7529
```
Comment (rossabaker) [inline build.sbt]: I know this gives us smaller jars.  I don't know what other speedups it brings.  I don't fully understand the three flags, or how they interact.  Would appreciate some feedback here from someone knowledgeable.

With Scala 2.12 looming, might we get away with dropping these if the Java 8 backend is more mature in Scala 2.12?

File: build.sbt [hunk 2]
```diff
@@ -389,22 +391,33 @@ lazy val commonSettings = Seq(
...
     "-unchecked",
-    "-Xlint"
-  )},
+    "-Xfatal-warnings",
+    "-Xlint",
+    "-Yinline-warnings",
+    "-Yno-adapted-args",
+    "-Ywarn-dead-code",
+    "-Ywarn-numeric-widen",
+    "-Ywarn-value-discard",
```
Comment (rossabaker) [inline build.sbt]: This is the one that I felt led to the most busy work.

**Comment by rossabaker** [comment]
Note also that 2.11.8 is a little picker than 2.10.6 on warnings.  This might be the time to cut over to 2.11 as the default build and 2.10 as the one Travis provides a safety net for.  With the introduction of macrocompat, this should now be viable.

**Comment by rossabaker** [comment]
* This is the end of the line for json4s-native.  Is that okay?
* Need to allow deprecation for scalaz-7.2 to avoid a branch.

File: client/src/test/scala/org/http4s/client/middleware/FollowRedirectSpec.scala [hunk 3]
```diff
@@ -42,7 +42,7 @@ class FollowRedirectSpec extends Http4sSpec with Tables {
       def doIt(method: Method, status: Status, body: String, pure: Boolean, response: Throwable \/ RedirectResponse) = {
         val u = uri("http://localhost") / status.code.toString
         val req = method match {
-          case _: Method with Method.PermitsBody if body.nonEmpty =>
+          case _: Method.PermitsBody if body.nonEmpty =>
```
Comment (rossabaker) [inline client/src/test/scala/org/http4s/client/middleware/FollowRedirectSpec.scala]: Scala 2.11 complained of a fruitless type until `PermitsBody` extended `Method`. Even a self-type wasn't enough.  What's up with this?

**Comment by rossabaker** [comment]
codecov/patch is bad because I reindented InstrumentedHandler.  I'd like a pass on this one.

**Comment by rossabaker** [comment]
I'd like to get this merged if Travis likes it.  This is difficult to keep in sync with master.

File: blaze-core/src/test/scala/org/http4s/blaze/TestHead.scala [hunk 4]
```diff
@@ -35,6 +35,7 @@ abstract class TestHead(val name: String) extends HeadStage[ByteBuffer] {
     closed = true
     super.stageShutdown()
     p.trySuccess(ByteBuffer.wrap(getBytes()))
+    ()
```
Comment (bryce-anderson) [inline blaze-core/src/test/scala/org/http4s/blaze/TestHead.scala]: These are curious. What flag requires them?

**Comment by bryce-anderson** [comment]
Looks good to me, though it wasn't extremely thorough. Given the nature of the changes, I'm happy to merge it.

**Comment by bryce-anderson** [comment]
Sorry, by 'it' I meant my review. The work is really good, but sometimes my use of English is not. 😀

File: blaze-core/src/test/scala/org/http4s/blaze/TestHead.scala (hunk 4 above)
Comment (rossabaker) [inline blaze-core/src/test/scala/org/http4s/blaze/TestHead.scala]: -Ywarn-value-discard

Of the standard flags that all the cool kids use, it's the one that I'm the most skeptical of.  Especially now that procedure syntax is deprecated, and we don't get bitten by missing equals signs on functions with inferred return types.

Comment (rossabaker) [inline blaze-core/src/test/scala/org/http4s/blaze/TestHead.scala]: That said, `Future[Unit]` silently treated as `Unit` was a big pain in my Akka days. I think this is a valuable flag on balance.  But it is the one that left the most curious artifacts in this PR.

**Pull request merged**